FreeBSD 9.1 support + start daemons

geoffgarside opened the pull request:
The following commits add FreeBSD 9.1 support along side FreeBSD 9.0. They are both pretty much the same, all they do is provide the `_freebsd_90_` and `_freebsd_91_` methods which call a `_freebsd_9x_` function.

I've also added /etc/rc.conf enabling of the daemons as per the [FreeBSD Installation Instructions](http://docs.saltstack.org/en/latest/topics/installation/freebsd.html).

Another pass has been made on the DISTRO_NAME_L setting, FreeBSD doesn't seem to like the `sed -re 's/(\s)+/_/g'` method either though `sed -re 's/[[:space:]]+/_/g'` does seem to be working in both FreeBSD and Ubuntu for me.

I've changed the name of the "salt" FreeBSD package to "sysutils/py-salt" to reflect the new package name. You could also use "py27-salt" but the fully qualified port name has the advantage of supporting any future python version defaults used in the package repository.

As a potential fix for #20 I've added the ability for the pkgng PACKAGESITE variable to be set by the environment, I'm not too sure if the `VAR=${VAR:="default"}` syntax is plain POSIX or not though.

Apologies for this being a rather disjointed set of commits, if you'd like me to break them up a bit into more logical chunks I'll happily do so.

geoffgarside committed:
Add support for FreeBSD 9.1 as well as 9.0.

geoffgarside committed:
FreeBSD update salt package -> sysutils/py-salt.

The salt package was renamed to the py- prefix to match other python
based ports. When built the name will be given a shortname of py27-salt,
in order to avoid Python versioning issues in the future we use the full
portname "sysutils/py-salt" which should always find the appropriate
version.

geoffgarside committed:
Add start daemons support for FreeBSD.

geoffgarside committed:
Protect from spamming /etc/rc.conf repeatedly.

The _post functions can be run multiple times. Make sure we don't write
the same string into /etc/rc.conf more than once.

geoffgarside committed:
Fix DISTRO_NAME_L for FreeBSD.

I've tested this with Ubuntu as well as FreeBSD and the [[:space:]]
method appears to work successfully across both OS's.

geoffgarside committed:
Allow PACKAGESITE to be overridden by env variable.

If the PACKAGESITE environment variable is not set then the normal
pkgbeta URL will be generated, if the variable is set then this will be
used in preference to the pkgbeta URL.

s0undt3ch commented:
@geoffgarside, first of all, Thanks!

I haven't looked at all your changes, though, the `DISTRO_NAME_L`, could you please verify that it works and change it to `sed -re 's/([[:space:]])+/_/g'`. That way we collapse multiple whitespace into a single underscore:

```
$ echo 'Foo Bar and Foo  Bar   FreeBSD' | sed -re 's/([[:space:]])+/_/g'
Foo_Bar_and_Foo_Bar_FreeBSD
```

s0undt3ch commented:
About the `POSIX` or not `POSIX` issues, we now have a test-suite. One of the first tests it does is `POSIX` compatibility, have a look [here](https://github.com/saltstack/salt-bootstrap/tree/develop/tests), more specifically, [here](https://github.com/saltstack/salt-bootstrap/blob/develop/tests/bootstrap/test_lint.py)

s0undt3ch commented:
If `install_freebsd_90_stable_deps` and `install_freebsd_91_stable_deps` do the exact same thing, couldn't we just define `install_freebsd_stable_deps`? Or will that bring problems with FreeBSD < 9.0 ?

s0undt3ch commented:
Sorry if I've sounded picky so far :smile:

geoffgarside commented:
Yes the `echo 'Foo Bar and Foo  Bar   FreeBSD' | sed -re 's/([[:space:]])+/_/g'` works, I'll change it to use that. I'll get a VM to run the tests in as it says it'll install/remove packages.

I believe the pkgng system should be supported in FreeBSD 8.2+ and 9.0+, it's possible people might bootstrap older versions but I could add a check to make sure the major version isn't less than 8. I simply opted for the most similar method of adding FreeBSD 9.1 based on what was already in the bootstrap.

Not at all, picky is a very good thing, especially for a multi-platform bootstrap project.

geoffgarside commented:
I'd need to get the Major version anyway in order to work out the pkgng ABI value as FreeBSD 8 will use "freebsd:8:${BSD_ARCH}" while the code at the moment assumes "freebsd:9:${BSD_ARCH}" so adding the check to make sure its not less than 8 shouldn't be a problem.

s0undt3ch commented:
#37 already mentions major and minor distro version support. Need to get that included.

s0undt3ch commented:
We can leave the 9.0 and 9.1 support as is. Maybe if someone complains, we later add 8 support?

geoffgarside committed:
Use suggested form for DISTRO_NAME_L final sed.

geoffgarside committed:
Copy {master,minion}.sample files if enabled.

geoffgarside commented:
Yeah can do FreeBSD 8 later if theres a demand for it, hopefully after #37 :)

I've run the Lint tests successfully and the Usage ones on a FreeBSD VM, I'll have to look further into the Installation tests and probably refamiliarise myself with Python :P

geoffgarside commented:
Yes good point, they do seem to be reliably installed by the port at the moment, but that could change.

geoffgarside committed:
Make {master,minion}.sample config copying safer.

We make sure they exist first :)

s0undt3ch commented:
I showed you that code because it will only copy if the file exists and, if we're only installing a minion, the master file won't be copied over because it's skipped some lines above :)

s0undt3ch commented:
heh, I've read your changes only now :smile: 

A single:

```
[ -f /usr/local/etc/salt/${fname}.sample ] && cp /usr/local/etc/salt/${fname}.sample /usr/local/etc/salt/${fname}
```

is enough.

`[ $fname = "minion" ] &&` is not necessary since if we skip on the loop if not installing, see [here](https://github.com/geoffgarside/salt-bootstrap/blob/5545aa6cfdb162c83da2f1feeb8b5cb571bd98ac/bootstrap-salt.sh#L1378)

geoffgarside commented:
Ah yes, now I've got you, the additional tests for fname value are redundant.

geoffgarside committed:
Remove redundant $fname tests when copying configs.

Testing for the sample file existence will protect against the lack of a
syndic.sample file.

s0undt3ch commented:
Yes. Also, [here](https://github.com/geoffgarside/salt-bootstrap/blob/5545aa6cfdb162c83da2f1feeb8b5cb571bd98ac/bootstrap-salt.sh#L1389) you only take into account the minion paths. The master and syndic paths are not important in FreeBSD?

s0undt3ch commented:
You noticed that now your sample copy lines are equal right? ;)

geoffgarside commented:
Only the `salt_minion` rc.d script has any paths configuration options.

geoffgarside committed:
Remove duped lines now $fname checks are gone.

s0undt3ch commented:
Ok, if those path changes makes it all work, I'm fine with it :)

geoffgarside commented:
Oh dear, school boy errors, its been an exhausting day.

geoffgarside commented:
The `salt_minion_paths` are lifted straight from the [FreeBSD Install guide](http://docs.saltstack.org/en/latest/topics/installation/freebsd.html) and include all the main paths. I suspect its required as without a shell most processes wouldn't get the /usr/local/{bin,sbin} paths which is where ports are installed.

s0undt3ch commented:
Well, I can tell you that I'm feeling bad for throwing all this at you :)

These are normal mistakes, I have just been devoting more time to this script, hence me picking this issues up.

s0undt3ch commented:
Ah! I think that the paths entry is there in order for the necessary binaries are found by the minion when running.

geoffgarside commented:
No no, don't feel bad about it, good to know its well curated :)

I've been looking through the tests/bootstrap/ext/os_data.py, it looks like just the os_family, oscodename, osfullname and osrelease grains aren't being detected for FreeBSD. Other than those are there any others I should be looking to populate. I'm not sure what the stubbed `_bsd_cpudata(grains)` ought to be doing.

s0undt3ch commented on bootstrap-salt.sh:
```diff
@@ -1310,31 +1310,37 @@ install_freebsd_90_stable_deps() {
     fi
 
-    fetch http://pkgbeta.freebsd.org/freebsd:9:${BSD_ARCH}/latest/Latest/pkg.txz
+    PACKAGESITE=${PACKAGESITE:="http://pkgbeta.freebsd.org/freebsd:9:${BSD_ARCH}/latest"}
```
Last change :smile: 

You see any issues with making this line like:

```
PACKAGESITE=${BS_PACKAGESITE:="http://pkgbeta.freebsd.org/freebsd:9:${BSD_ARCH}/latest"}
```

I've prefixed the environment variable with `BS_`.

geoffgarside replied:
Actually I think it probably needs to be

`BS_PACKAGESITE=${PACKAGESITE:-"http://pkgbeta.freebsd.org/freebsd:9:${BSD_ARCH}/latest"}`

so the `:-` format won't set $PACKAGESITE but if its unset $BS_PACKAGESITE will get the value or that of $PACKAGESITE if its defined. I'd probably try to keep $PACKAGESITE as the environment variable as thats in keeping with the environment variables you can use to override `pkg`.

s0undt3ch replied:
I am using `:-` [here](https://github.com/saltstack/salt-bootstrap/blob/develop/bootstrap-salt.sh#L41) and it does work.

Regarding the `PACKAGESITE` env variable, if FreeBSD already uses it, ok, let's not rename it. This is valuable information(because I would have renamed it :smiley:)

geoffgarside replied:
Its probably better form to check for the existence of environment variables and give them internal names for continued use and setting defaults and the like though. I'll keep the environment variable as PACKAGESITE and internally store that (or use the determined default) as BS_PACKAGESITE.

s0undt3ch commented:
The code in `tests/bootstrap/ext/os_data.py` was taken [directly from salt ](https://github.com/saltstack/salt/blob/develop/salt/grains/core.py#L548). I probably did not copy enough code for FreeBSD to be properly detected.

geoffgarside committed:
Store PACKAGESITE in BS_PACKAGESITE.

If the PACKAGESITE environment variable is set then we will use that
value, otherwise we calculate a sensible default and use that instead.

geoffgarside commented:
I think I need to read through the tests and work out how much of the grain information is actually used before I go and copy the `_bsd_cpudata()` method across :-P

s0undt3ch commented:
Currently we're using the grains so we know how to un-install any salt packages and their dependencies, [have a look here](https://github.com/geoffgarside/salt-bootstrap/blob/5545aa6cfdb162c83da2f1feeb8b5cb571bd98ac/tests/bootstrap/test_install.py#L17)

s0undt3ch commented:
So, right now, only OS Family.

geoffgarside committed:
Add FreeBSD detection to bootstrap.ext.os_data

geoffgarside committed:
Add FreeBSD test_install cleanup rules.

To the `pkg autoremove` command will remove any packages from the system
which are not currently depended on by any other packages.

s0undt3ch commented:
WOW! You even added the required code for the tests! A BIG THANK YOU!

geoffgarside commented:
Nuts, the tests aren't working yet :(

s0undt3ch commented:
You're running?

```
PACKAGESITE=foo sudo -E tests/runtests.py
```

s0undt3ch commented:
Or you might be running as root :)

geoffgarside commented:
Yeah, running as root on a VM and I've got my PACKAGESITE set. It might be failing on the cleanup

```
ERunning cleanup command 'pkg delete -y sysutils/py-salt'
======================================================================
ERROR: test_install_daily (bootstrap.test_install.InstallationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/salt-bootstrap/tests/bootstrap/test_install.py", line 139, in test_install_daily
    args=('daily',), timeout=15 * 60, stream_stds=True
  File "/root/salt-bootstrap/tests/bootstrap/__init__.py", line 138, in run_script
    while process.poll() is None:
  File "/root/salt-bootstrap/tests/bootstrap/__init__.py", line 74, in poll
    obuff = self.stdout.read()
IOError: [Errno 35] Resource temporarily unavailable
```

s0undt3ch commented:
I'm only taking into account errno 11 in the tests, that might be it.

Looking [here](http://fxr.watson.org/fxr/source/sys/errno.h#L92) the fix should be, in [line 78'](https://github.com/saltstack/salt-bootstrap/blob/develop/tests/bootstrap/__init__.py#L78) and [line 88](https://github.com/saltstack/salt-bootstrap/blob/develop/tests/bootstrap/__init__.py#L88) of [tests/bootstrap/**init**.py](https://github.com/saltstack/salt-bootstrap/blob/develop/tests/bootstrap/__init__.py):

``` python
            except IOError, err:
                if err.errno not in (11, 35):
```

geoffgarside commented:
Thanks, thats worked, two errors now, one in test_install_daily and test_install_using_bash, I'm guessing some part of the cleanup is failing. I'm re-running the tests now and capturing the output to see which cleanup command it running and then getting an "E".

```
======================================================================
ERROR: test_install_daily (bootstrap.test_install.InstallationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/salt-bootstrap/tests/bootstrap/test_install.py", line 55, in tearDown
    stream_stds=True
  File "/root/salt-bootstrap/tests/bootstrap/__init__.py", line 191, in assert_script_result
    err_msg = '{0} No std{out,err} captured.'.format(err_msg)
KeyError: 'out,err'

======================================================================
ERROR: test_install_using_bash (bootstrap.test_install.InstallationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/salt-bootstrap/tests/bootstrap/test_install.py", line 55, in tearDown
    stream_stds=True
  File "/root/salt-bootstrap/tests/bootstrap/__init__.py", line 191, in assert_script_result
    err_msg = '{0} No std{out,err} captured.'.format(err_msg)
KeyError: 'out,err'

----------------------------------------------------------------------
```

s0undt3ch commented:
Ops, that's also my fault.

Change:

``` python
err_msg = '{0} No std{out,err} captured.'.format(err_msg)
```

to

``` python
err_msg = '{0} No stdout nor stderr captured.'.format(err_msg)
```

s0undt3ch commented:
But apparently an error did occur, we just did not get any output

geoffgarside commented:
Ok, applied the change, running tests again. So is Python wanting to interpolate into the `{out,err}` bit then? My python skills are rather rusty.

geoffgarside commented:
Ok, got an error message :)

```
AssertionError: Failed to execute cleanup command 'pkg delete -y swig sysutils/py-salt':
 No stdout nor stderr captured.
```

though I've no idea why that wouldn't return any stdout or stderr, running them manually you get output whether they're installed or not.

s0undt3ch commented:
About the interpolation, yes, `{0}` are list indexes, `{foo}` is dictionary keys using `.format()` on strings.

Similiar to

``` python
'%s %(foo)s' % 'a' % dict(foo=b)
```

s0undt3ch commented:
What's the exit code when you run:

```
'pkg delete -y swig sysutils/py-salt'
```

ie:

```
echo $?
```

geoffgarside committed:
Add swig to list of test_install cleanup items.

geoffgarside commented:
When the packages are installed and you run that, you get 0.
When the packages are missing and you run that, you get 65.

I think the 65 corresponds to the pkgng EPKG_\* codes.

Thanks for the info on the interpolation :)

s0undt3ch commented:
Ok, got another fix for you:

``` diff
diff --git a/tests/bootstrap/__init__.py b/tests/bootstrap/__init__.py
index ff26c15..0e02d7c 100644
--- a/tests/bootstrap/__init__.py
+++ b/tests/bootstrap/__init__.py
@@ -179,9 +179,12 @@ class BootstrapTestCase(TestCase):
                 # process already terminated
                 pass

-    def assert_script_result(self, fail_msg, expected_rc, process_details):
+    def assert_script_result(self, fail_msg, expected_rcs, process_details):
+        if not isinstance(expected_rcs, (tuple, list)):
+            expected_rcs = (expected_rcs,)
+
         rc, out, err = process_details
-        if rc != expected_rc:
+        if rc not in expected_rcs:
             err_msg = '{0}:\n'.format(fail_msg)
             if out:
                 err_msg = '{0}STDOUT:\n{1}\n'.format(err_msg, '\n'.join(out))
diff --git a/tests/bootstrap/test_install.py b/tests/bootstrap/test_install.py
index 096f506..6631488 100644
--- a/tests/bootstrap/test_install.py
+++ b/tests/bootstrap/test_install.py
@@ -43,7 +43,11 @@ class InstallationTestCase(BootstrapTestCase):
             print 'Running cleanup command {0!r}'.format(cleanup)
             self.assert_script_result(
                 'Failed to execute cleanup command {0!r}'.format(cleanup),
-                0,
+                (
+                    0,
+                    65  # FreeBSD throws this error code when the packages being
+                        # un-installed were not installed in the first place
+                ),
                 self.run_script(
                     script=None,
                     args=cleanup.split(),
```

geoffgarside commented:
Yay

```
Ran 8 tests in 176.167s

OK (skipped=1)
```

I suspect the bash one might have been skipped as /bin/bash doesn't exist on FreeBSD, ends up in /usr/local/bin/bash once the package is installed.

s0undt3ch commented:
That is correct!

Can I add your name and email to the script's header?

s0undt3ch commented:
You can also add it if you like :)

geoffgarside commented:
Shall I apply your fixes in my branch or do you want to apply them to master and I'll merge master into my branch? Yeah I can add that to the header :)

s0undt3ch commented:
Nah! You can add all fixes!

geoffgarside committed:
Add @s0undt3ch provided fixes, many thanks.

geoffgarside committed:
Add name to header.

geoffgarside commented:
Well, after 3 hours of bug squashing and very kind assistance I think it might just be ready :)

s0undt3ch commented:
Well, I'm just finishing some more tweaks to the now included #37, which will allow future FreeBSD 8 support, if anyone asks for it, and I'll merge your changes.

Once again. Thank You!

geoffgarside commented:
Thats cool, by the looks of things, the simplest way to add FreeBSD 8 support would be to put `$DISTRO_MAJOR_VERSION` in the `BS_PACKAGESITE` variable value and then berate anyone who tried to do it on any earlier FreeBSD versions :P

s0undt3ch commented:
Yep!

s0undt3ch commented:
Ok, could you please pull latest changes and see if it still works?

geoffgarside commented:
Ok, will let you know shortly.

s0undt3ch commented:
There was an additional fix to `DISTRO_MINOR_VERSION`.

geoffgarside commented:
Yeah noticed some sed errors, fetched and running testsuite again :)

geoffgarside commented:
Yep, all tests passing, 1 skipped

s0undt3ch commented:
Awesome!!!!

Once again, Thanks!!!

## 

Pedro Algarvio

----- Reply message -----
From: "Geoff Garside" notifications@github.com
To: "saltstack/salt-bootstrap" salt-bootstrap@noreply.github.com
Cc: "Pedro Algarvio" pedro@algarvio.me
Subject: [salt-bootstrap] FreeBSD 9.1 support + start daemons (#47)
Date: Tue, Feb 12, 2013 21:38
Yep, all tests passing, 1 skipped 

&mdash;

Reply to this email directly or view it on GitHub.
